File: src/TheWorld/wwwroot/js/tripsController.js
/// <reference path="../lib/angular/angular.js" />
/// <reference path="../lib/angular/angular.min.js" />

(function () {
    "use strict";
    angular.module("app-trips")
    .controller("tripsController", tripsController);

    function tripsController($http) {
        var vm = this;

        vm.trips = [];

        vm.newTrip = {};
        vm.errorMessage = "";
        vm.isBusy = true;
        $http.get("/api/trips")
        .then(function (response) {
            //success
            angular.copy(response.data, vm.trips);

        },function(error){
            vm.errorMessage = "Failed to load data: " + error;        
        })
        .finally(function() {
            vm.isBusy = false;
            vm.errorMessage = "";
            });
        vm.addTrip = function () {
            vm.isBusy = true;
            $http.post("/api/trips", vm.newTrip)
            .then(function (response) {
                //Success
                vm.trips.push(response.data);
                vm.newTrip = {};
            }, function (response) {
                //failure
                vm.errorMessage = "Failed to save new trip";
            })
            .finally(function () {
                vm.isBusy = false;
            });

        };

    }
})();

File: src/TheWorld/Controllers/Api/TripsController.cs
﻿using AutoMapper;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.Extensions.Logging;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using TheWorld.Models;
using TheWorld.ViewModels;

namespace TheWorld.Controllers.Api
{
    [Route("api/trips")]
    [Authorize]
    public class TripsController : Controller
    {
        private ILogger<TripsController> _logger;
        private IWorldRepository _repository;

        public TripsController(IWorldRepository repository, ILogger<TripsController> logger)
        {
            _repository = repository;
            _logger = logger;
        }


        [HttpGet("")]
        public IActionResult Get()
        {
            try
            {
                var results = _repository.GetTripsByUsername(this.User.Identity.Name);
                return Ok(Mapper.Map<IEnumerable<TripViewModel>>(results));
            }catch(Exception ex)
            {
                _logger.LogError($"Failed to get All Trips: {ex}");
                return BadRequest("Error Occured");
            }
            
        }

        [HttpPost("")]
        public async Task<IActionResult> Post([FromBody]TripViewModel trip)
        {
            if (ModelState.IsValid)
            {
                //Save to the Database
                var newTrip = Mapper.Map<Trip>(trip);
                newTrip.UserName = User.Identity.Name;

                _repository.AddTrip(newTrip);

                if(await _repository.SaveChangesAsync())
                {
                    return Created($"api/trips/{trip.Name}", Mapper.Map<TripViewModel>(newTrip));
                }                
            }
            return BadRequest("Failed to save the Trip");
        }
    }
}
